### DISCUSSION TITLE
Can't have default values for dynamic forms

### DISCUSSION THREAD

**Discussion by eakl** [react-hook-form/react-hook-form#6063] [Q&A] [answered]
Hi guys,

First of all, here is the CSB: https://codesandbox.io/s/immutable-field-z372p?file=/src/App.js

### What I am trying to achieve
I have:
- A main form
- that contains multiple subforms which will be mounted as soon as a particular field (an option in a select input) is selected
- A summary table that displays an array of metrics

When the `type` field changes, the main form renders the associated subform and the summary table displays the `metrics` field.

### What is wrong with my approach
Only the main form has `defaultValues` because the subforms are dynamic upon user selection (and I may add more subform overtime) and I don't know which subform should be in the final form data when submitted by the user.

However, The `Table` component must always display `defaultValues` for all the metrics of the subform as soon as the subform is mounted. Then, it should unmount the previous `defaultValues` and mount the new ones when another subform is selected.

e.g. of What I would like to achieve

```javascript
// Step 1: You load the main form
metrics: []

// Step 2: You select "Sub form 1"
metrics: [{ value: 001 }, { value: 002 }] // both fields should appear with their defaults even tho the fields haven't be touched yet (isDirty: false??)

// Step 3: You type '23' in input 001 of "Sub form 1"
metrics: [{ value: 23 }, { value: 002 }] // input 001 has the value '23' and input 002 has it's default value (even tho is hasn't been touched)

// Step 4: You select "Sub form 2"
metrics: [{ value: 003 }, { value: 004 }] // previous values of the `sub form 1` should disappear and be replaces with default values of `sub form 2`
```

Even better solution (which is what I need for my project) is to have an object as default values 

```javascript
// Step 1: You load the main form
metrics: []

// Step 2: You select "Sub form 1"
metrics: [
  {
    form: 'subform1',
    input: 001,
    value: 001
  },
  {
    form: 'subform1',
    input: 002,
    value: 002
  }
] // both fields should appear with their default objects even tho the fields haven't be touched yet (isDirty: false??)

// Step 3: You type '23' in input 001 of "Sub form 1"
metrics: [
  {
    form: 'subform1',
    input: 001,
    value: 23
  },
  {
    form: 'subform1',
    input: 002,
    value: 002
  }
] // input 001 has its default object and the key `{ value: '23' }` and input 002 has it's default object (even tho is hasn't been touched)

// Step 4: You select "Sub form 2"
metrics: [
  {
    form: 'subform1',
    input: 003,
    value: 003
  },
  {
    form: 'subform1',
    input: 004,
    value: 004
  }
] // previous values of the `sub form 1` should disappear and be replaces with default values of `sub form 2`

// Step 5: You submit the form
formData: {
  name: 'default name',
  type: 'subform2',
  metrics: [
    {
      form: 'subform1',
      input: 003,
      value: 003
    },
    {
      form: 'subform1',
      input: 004,
      value: 004
    }
  ]
}
```

Any guidance on how I can achieve this with react-hook-form?

Thanks a lot

**Comment by bluebill1049** [comment, marked as answer]
- If you have provided `defaultValues` then when inputs get mounted, that input should be populated with `defaultValue`
- if your data is dynamic use `reset` https://react-hook-form.com/api/useform/reset
- if you want to update single or couple inputs you can use `setValue` https://react-hook-form.com/api/useform/setvalue

**Reply by eakl** [reply]
Thanks @bluebill1049 for your answer.

I have a question about your previous message:

**NOTE: I have updated the CSB**

### How `defaultValue` works with `defaultValues`?

I can't provide `defaultValues` for all subforms at the parent form component (in `useForm()`) as the parent component shouldn't know about the subforms. So `defaultValues` should be set at the subform level

However the following doesn't work

```javascript
// Parent form
export default function App() {
  const form = useForm({
    defaultValues: { name: "default name", type: "none", metrics: [] }
  });
}

// Subform
<InputNumber
  className="elem"
  name="metrics.1.value"
  placeholder="002"
  control={form.control}
  defaultValue={{ type: "number", value: "002" }} // set defaultValues at the subform level
/>

// The Input
export const InputNumber = ({ name, control, defaultValue, ...rest }) => {
  const {
    field: { ref, ...field }
  } = useController({
    name,
    control,
    defaultValue // setting the default values here
  });

  return <input type="number" {...rest} {...field} />;
};
```

What I can do is to `setValue` in a useEffect inside the subform like so

```javascript
// Subform1 Component
useEffect(() => {
    form.setValue("metrics", [
      { type: "number", value: "003" },
      { type: "number", value: "004" }
    ]);
  }, []);
```

But then, when the select input is changed back form `Sub form 1` to `none`, the default values are not unregistered completely. Only the `value` key is unregistered

**Reply by gumpenaajayvarma** [reply]
You can create a HOC () in which you will be doing the useForm with default values or other options.
Use the HOC for each sub form so that each sub form is isolated to its own hook form.

```
export function withHookFormProvider (WrappedComponent, componentProps, formProps) {
  const HookForm = () => {
    const formAPI = useForm(formProps)
    return (
      <FormProvider {...formAPI}>
        <form>
          <WrappedComponent {...componentProps} />
        </form>
      </FormProvider>
    )
  }
  HookForm.displayName = WrappedComponent.displayName || WrappedComponent.name || 'HookForm'
  return HookForm
}
```

**Reply by eakl** [reply]
@gumpenaajayvarma I need to submit all values (parent form AND subform) in the same payload when hitting "Submit" in the Parent form.

I wish I can register and unregister additional fields on the fly when changing the select field of the main form. Except that the field I register should be an object with a bunch of defaults and the `value` key that will change as the field input changes.

Basically the when the main form is loaded, the form state looks like this
```javascript
{
  name: 'Name of the main form',
  type: 'none',
  metrics: []
}
```

When you change the `type` field to `Sub Form 1` a subform with 3 inputs is loaded and the sate looks like this
```javascript
{
  name: 'Name of the main form',
  type: 'subform1', // here you have the subform type that comes with default metrics
  metrics: [
    { subform: 'Sub Form 1', inputNumber: 001, value: '333' }, // 'value' will change on Change 
    { subform: 'Sub Form 1', inputNumber: 002, value: '666' }, // 'value' will change on Change
    { subform: 'Sub Form 1', inputNumber: 003, value: '999' } // 'value' will change on Change
  ]
}
```

when you select back 'none' on the `type` field, you get back to this
```javascript
{
  name: 'Name of the main form',
  type: 'none',
  metrics: []
}
```

After @bluebill1049 comment, I can set the default metrics values in a useEffect hook when the subform is mounted but I can't clear them when the subform is unmounted, even with a cleanup function

```javascript
useEffect(() => {
  form.setValue("metrics", [
    { subform: 'Sub Form 1', inputNumber: 002, value: '333' },
    { subform: 'Sub Form 1', inputNumber: 002, value: '666' },
    { subform: 'Sub Form 1', inputNumber: 002, value: '999' },
  ]);

  return () => {
    form.setValue('metrics', []) // Doesn't work
  }
}, []);
```
